File: 0072-edit-distance/0072-edit-distance.cpp
class Solution {
public:
    int min3(int a, int b, int c) {
        return min(a, min(b, c));
    }
    
    int minDistance(string word1, string word2) {
        int n = (int) word1.size();
        int m = (int) word2.size();
        vector<vector<int>> dp(n+1, vector<int>(m+1));
        for(int i = 0; i <= n; i++) {
            for(int j = 0; j <= m; j++) {
                if(i == 0) dp[i][j] = j;
                else if(j == 0) dp[i][j] = i;
                else if(word1[i-1] == word2[j-1]) dp[i][j] = dp[i-1][j-1];
                else {
                    dp[i][j] = 1 + min3(dp[i-1][j-1], dp[i][j-1], dp[i-1][j]);
                }
            }
        }
        return dp[n][m];
    }
};

File: 0072-edit-distance/0072-edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n,m = len(word1), len(word2)
        dp = [[-1 for j in range(m+1)] for i in range(n+1)]
        for i in range(0,n+1):
            dp[i][m] = n-i
        for j in range(0,m+1):
            dp[n][j] = m-j
        for i in range(n-1,-1,-1):
            for j in range(m-1,-1,-1):
                if word1[i] == word2[j] :
                    dp[i][j] = dp[i+1][j+1]
                else:
                    dp[i][j] = 1 + min(min(dp[i+1][j], dp[i][j+1]), dp[i+1][j+1])
        return dp[0][0]